lib/web_ui/lib/src/engine/semantics/semantics.dart

```diff
@@ -1513,20 +1552,30 @@ class EngineSemanticsOwner {
   /// the one-time callbacks scheduled via the [addOneTimePostUpdateCallback]
   /// method.
   void _finalizeTree() {
-    for (final SemanticsObject? object in _detachments) {
-      final SemanticsObject? parent = _attachments[object!.id];
-      if (parent == null) {
-        // Was not reparented and is removed permanently from the tree.
-        _semanticsTree.remove(object.id);
-        object._parent = null;
-        object.element.remove();
-      } else {
-        assert(object._parent == parent);
-        assert(object.element.parentNode == parent._childContainerElement);
+    for (final SemanticsObject detachmentRoot in _detachments) {
+      // A detached node may or may not have some of its descendants reattached
+      // elsewhere. Walk the descendant tree and find all descendants that were
+      // reattached to a parent. Those descendants need to be removed.
+      final List<SemanticsObject> removals = <SemanticsObject>[];
+      detachmentRoot.visitDepthFirst((SemanticsObject node) {
```

Comment:
This is traversing the entire tree including all descendants of detached nodes. Do you think it's worth making a short circuit to avoid traversing children of detached nodes?

It can be done by returning a boolean from the callback to indicate whether you want to continue down that subtree or not.

Reply:
See https://github.com/flutter/engine/pull/41681#discussion_r1189263649